### PR TITLE
Per-channel scroll behaviour

### PR DESCRIPTION
**Pull request by notbakaneko**
Handles scroll position on a per-channel basis.
Also fixes the initial scroll not working properly when navigating to the page.

- window focus changed to an `observable`
- mark as read now only happens when scroll is at the bottom of a channel when:
  - switching away from the channel to another channel (so scrolling up and then switching will preserve the marker)
  - window comes into focus
  - sending a message (which still auto scrolls to the bottom)

More accurate handling of marking as read needs a bit more complicated tracking of the position of the unread marker which I really don't want to do here 👀

### CODE REVIEW

**Comment by nanaya** [comment]
conflicts...

File: resources/assets/lib/chat/conversation-view.tsx [hunk 1]
```diff
@@ -116,23 +127,33 @@ export default class ConversationView extends React.Component<Props> {
...
     if (this.didSwitchChannel) {
-      if (this.unreadMarkerRef.current) {
-        this.scrollToUnread();
-      } else {
-        this.scrollToBottom();
-      }
-      this.didSwitchChannel = false;
+      // This can happen after a turbolinks navigation,
+      // so we have to wait for the elements to be on the current document before scrolling.
+      core.reactTurbolinks.runAfterPageLoad(this.id, action(() => {
```
Comment (nanaya) [inline resources/assets/lib/chat/conversation-view.tsx:135]: also need listener removal for this

Looking again the `runAfterPageLoad` function should just return a disposer :thinking: then the id wouldn't be needed anymore. And looking even closer I noticed one is called synchronously and the other is not - that may cause confusing behavior difference.

File: resources/assets/lib/chat/conversation-view.tsx [hunk 2]
```diff
@@ -94,6 +96,15 @@ export default class ConversationView extends React.Component<Props> {
 
     makeObservable(this);
 
+    disposeOnUnmount(
+      this,
+      observe(core.windowFocusObserver, 'hasFocus', (change) => {
```
Comment (nanaya) [inline resources/assets/lib/chat/conversation-view.tsx:101]: just a minor thing but according to the docs [it's preferred to use reaction over observe](https://mobx.js.org/intercept-and-observe.html#intercept--observe-) due to how observe works (ignores action and just run right away)

File: resources/assets/lib/chat/input-box.tsx [hunk 3]
```diff
@@ -40,10 +37,17 @@ export default class InputBox extends React.Component<Props> implements Dispatch
...
 
-    dispatcher.register(this);
-
     makeObservable(this);
 
+    disposeOnUnmount(
+      this,
+      autorun(() => {
+        if (core.windowFocusObserver.hasFocus) {
+          this.focusInput();
```
Comment (nanaya) [inline resources/assets/lib/chat/input-box.tsx:46]: it has always been the case but I just noticed opening quick search when on chat page and then switching focus out and back in results in being able to type in chat while quick search is still open :upside_down_face:

**Review by nanaya — CHANGES_REQUESTED** [review]
conflicts again...

File: resources/assets/lib/chat/conversation-view.tsx (hunk 1 above)
Comment (notbakaneko) [reply resources/assets/lib/chat/conversation-view.tsx:135]: doesn't `.one()` unbind automatically after it's done? 🤔

Comment (nanaya) [reply resources/assets/lib/chat/conversation-view.tsx:136]: that's assuming the component is still mounted when it runs

Comment (notbakaneko) [reply resources/assets/lib/chat/conversation-view.tsx:136]: sounds like `runAfterPageLoad` should change; `$(document).off` in the component without `$(document).on` seems counter intuitive 🤔

File: resources/assets/lib/chat/input-box.tsx [hunk 4]
```diff
@@ -73,11 +77,12 @@ export default class InputBox extends React.Component<Props> implements Dispatch
     this.focusInput();
   }
 
-  componentWillUnmount() {
-    dispatcher.unregister(this);
-  }
+  focusInput(always = true) {
+    const activeElement = document.activeElement;
+    if (!always && (activeElement instanceof HTMLInputElement || activeElement instanceof HTMLTextAreaElement)) {
```
Comment (nanaya) [inline resources/assets/lib/chat/input-box.tsx:82]: that's an attempt but there's nothing forcing focus on quick search input at the moment :eyes: 

(search thing -> open result in new tab with mouse (middle click etc) -> switch to the tab -> go back to chat tab)

**Review by nanaya — CHANGES_REQUESTED** [review]

**Review by nanaya — APPROVED** [review]

**Pull request merged**